### DISCUSSION TITLE
How to get access to KeyEventType

### DISCUSSION THREAD

**Discussion by quendimax** [slint-ui/slint#3352] [Q&A] [answered]
Hello,

I can't find a way to get access to `KeyEventType` from my rust source code. I want to check a value of `KeyEvent.event_type` member, but I can't. I have found that the enum resides in `i_slint_core::input` module, but I don't have access to it, and it is not re-exported via `slint` module.

**Comment by ogoffart** [comment]
This enum is currently internal (private API). `KeyEvent` is also private API.
Why do you need to have access to it?

**Reply by quendimax** [reply]
>  Why do you need to have access to it?

Actually there is no one clear answer. I'm writing a keybinding system that is configurable (using json configs), and it can contain sequences of keystrokes. Implementation all of this via Slint code looks too difficult now (if not impossible) because of the Slint's script language is not comprehensive now (doesn't have local variable, match/switch-case statements, enums, etc). So I decided to implement it on rust side, and it is why I use `KeyEvent` directly.

> This enum is currently internal (private API). `KeyEvent` is also private API.

Yes, now I use the following trick :
```
// in my `ui` module
pub type Key = slint::private_unstable_api::re_exports::Key;
pub type KeyEvent = slint::private_unstable_api::re_exports::KeyEvent;
pub type EventResult = slint::private_unstable_api::re_exports::EventResult;
```
and I thought I could use the same trick for `KeyEventType` struct. But it is not accessible via this module path. I was a bit confused, because I thought that it is not possible to use `KeyEvent::event_type` member if its type is private. But at the same time I couldn't find a way to get the `KeyEventType` directly.

**Reply by ogoffart** [reply, marked as answer]
Note that as the name says, this API is unstable and may change accross releases.

But you can forward a callback from a FocusScope
```
enum KeyEventAction { pressed, release }
//...
FocusScope {
    key-pressed(event) => {  SomeGlobal.key-event(KeyEventAction.pressed, event.text)  }
}
```

something similar to that.
And then handle these events in rust.
